# Tasks<br> 

<b>Part 1: </b><br>
1. Download the customer data set and add it to your “Original Data” folder.<br>
2. Create a new notebook in your “Scripts” folder for part 1 of this task.<br>
3. Import your analysis libraries, as well as your new customer data set as a dataframe.<br>
4. Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.<br>
5. Complete the fundamental data quality and consistency checks (check for and address missing values and duplicates, and convert any mixed-type data).<br>
6. Combine your customer data with the rest of your prepared Instacart data.<br>
7. Ensure your notebook contains logical titles, section headings, and descriptive code comments.<br>
8. Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.<br>
9. Save your notebook so that you can send it to your tutor for review after completing part 2.<br><br>

<b>Steps 1 and 2 have been completed. Onwards to Step 3. </b>

# Step 3. Import libraries and dataframe

In [1]:
# Import libraries 
import pandas as pd 
import numpy as np 
import os 
import matplotlib.pyplot as plt 
import seaborn as sns 
import scipy

In [2]:
# Create path 
path = r'/Users/bentley/Documents/Instacart'

In [5]:
# Import dataframe 
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

# Step 4. Data wrangling 

In [6]:
# Check the size of the dataframe 
df_customers.shape

(206209, 10)

In [8]:
# View the dataframe
df_customers.head(20)

user_id First Name    Surnam  Gender           STATE  Age date_joined  \
0     26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1     33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2     65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3    125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4    130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5    133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6    152052      Chris    Walton    Male         Montana   20    1/1/2017   
7    168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8     69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9     82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10    26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11    33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12    75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13       51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14      516      Peter      Hunt    Male        Colorado   51    1/1/2017   
15     6965    Brandon      Mata    Male    North Dakota   70    1/1/2017   
16     8794        Amy      Mack  Female        Nebraska   30    1/1/2017   
17     9944      Brian    Oneill    Male       Wisconsin   26    1/1/2017   
18    13313     Victor      Moss    Male          Alaska   31    1/1/2017   
19    17335    Patrick    Fowler    Male         Vermont   65    1/1/2017   

    n_dependants                        fam_status  income  
0              3                           married  165665  
1              0                            single   59285  
2              2                           married   99568  
3              0                            single   42049  
4              1                           married   40374  
5              2                           married   49643  
6              0                            single   61746  
7              0                            single   63712  
8              1                           married  162432  
9              2                           married   32072  
10             2  living with parents and siblings   43443  
11             2                           married   47988  
12             1                           married   57835  
13             1                           married   71190  
14             2                           married  146559  
15             0                  divorced/widowed  100091  
16             3                           married   57096  
17             1                           married   91826  
18             2                           married   91111  
19             1                           married   34514

In [9]:
# Rename a column 
df_customers.rename(columns = {'First Name' : 'first_name'}, inplace = True)

In [10]:
# Rename a column 
df_customers.rename(columns = {'Surnam' : 'last_name'}, inplace = True)

In [11]:
# Rename a column 
df_customers.rename(columns = {'Gender' : 'gender'}, inplace = True)

In [12]:
# Rename a column 
df_customers.rename(columns = {'Age' : 'age'}, inplace = True)

In [20]:
# Rename a column 
df_customers.rename(columns = {'n_dependants' : 'nbr_of_dependents'}, inplace = True)

In [14]:
# View dataframe 
df_customers.head()

user_id first_name last_name  gender       STATE  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [16]:
# Rename a column 
df_customers.rename(columns = {'STATE' : 'state'}, inplace = True)

In [17]:
# View dataframe 
df_customers.head()

user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

Names of columns have been checked and updated. 

In [18]:
# Check descriptive stats 
df_customers.describe()

user_id            age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [19]:
# Change datatype 
df_customers['user_id'] = df_customers['user_id'].astype('str')

In [22]:
# Check datatype 
df_customers['user_id'].dtype

dtype('O')

Records in 'user_id' column are objects (aka strings). 

In [23]:
# View dataframe 
df_customers.head(25)

user_id first_name last_name  gender           state  age date_joined  \
0    26711    Deborah  Esquivel  Female        Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female      New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male           Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female            Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female        Maryland   26    1/1/2017   
5   133128    Cynthia     Noble  Female        Kentucky   43    1/1/2017   
6   152052      Chris    Walton    Male         Montana   20    1/1/2017   
7   168851     Joseph   Hickman    Male  South Carolina   30    1/1/2017   
8    69965     Jeremy      Vang    Male           Texas   47    1/1/2017   
9    82820      Shawn     Chung    Male        Virginia   26    1/1/2017   
10   26441     Gloria  Stafford  Female          Nevada   19    1/1/2017   
11   33835      Roger     Casey    Male        Nebraska   45    1/1/2017   
12   75032    Cynthia  Mcdaniel  Female         Georgia   75    1/1/2017   
13      51  Stephanie    Jensen  Female         Wyoming   37    1/1/2017   
14     516      Peter      Hunt    Male        Colorado   51    1/1/2017   
15    6965    Brandon      Mata    Male    North Dakota   70    1/1/2017   
16    8794        Amy      Mack  Female        Nebraska   30    1/1/2017   
17    9944      Brian    Oneill    Male       Wisconsin   26    1/1/2017   
18   13313     Victor      Moss    Male          Alaska   31    1/1/2017   
19   17335    Patrick    Fowler    Male         Vermont   65    1/1/2017   
20   29864      Randy     Mills    Male    North Dakota   39    1/1/2017   
21   37744      Wanda     Salas  Female        Arkansas   77    1/1/2017   
22   50000     Sandra     Trejo  Female           Maine   64    1/1/2017   
23   55567     Martha     Hurst  Female  North Carolina   67    1/1/2017   
24   57549      Shawn     Moore    Male         Montana   72    1/1/2017   

    nbr_of_dependents                        fam_status  income  
0                   3                           married  165665  
1                   0                            single   59285  
2                   2                           married   99568  
3                   0                            single   42049  
4                   1                           married   40374  
5                   2                           married   49643  
6                   0                            single   61746  
7                   0                            single   63712  
8                   1                           married  162432  
9                   2                           married   32072  
10                  2  living with parents and siblings   43443  
11                  2                           married   47988  
12                  1                           married   57835  
13                  1                           married   71190  
14                  2                           married  146559  
15                  0                  divorced/widowed  100091  
16                  3                           married   57096  
17                  1                           married   91826  
18                  2                           married   91111  
19                  1                           married   34514  
20                  1                           married   44364  
21                  1                           married  125977  
22                  2                           married  115489  
23                  2                           married   48645  
24                  0                  divorced/widowed  135302

In [24]:
# View information 
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   user_id            206209 non-null  object
 1   first_name         194950 non-null  object
 2   last_name          206209 non-null  object
 3   gender             206209 non-null  object
 4   state              206209 non-null  object
 5   age                206209 non-null  int64 
 6   date_joined        206209 non-null  object
 7   nbr_of_dependents  206209 non-null  int64 
 8   fam_status         206209 non-null  object
 9   income             206209 non-null  int64 
dtypes: int64(3), object(7)
memory usage: 15.7+ MB


In [25]:
# Check frequency of states in the 'state' column 
df_customers['state'].value_counts()

Iowa                    4044
District of Columbia    4044
Colorado                4044
California              4044
Florida                 4044
Arkansas                4044
Connecticut             4044
Indiana                 4044
Alabama                 4044
Arizona                 4044
Illinois                4044
Georgia                 4044
Alaska                  4044
Hawaii                  4044
Delaware                4044
Idaho                   4044
Maine                   4043
Michigan                4043
Virginia                4043
Oregon                  4043
Wisconsin               4043
South Carolina          4043
New Mexico              4043
Vermont                 4043
New Jersey              4043
Oklahoma                4043
Maryland                4043
Montana                 4043
West Virginia           4043
Massachusetts           4043
Rhode Island            4043
New York                4043
Tennessee               4043
North Dakota            4043
Missouri      

In [26]:
# Sort states in indexing order 
df_customers['state'].value_counts().sort_index()

Alabama                 4044
Alaska                  4044
Arizona                 4044
Arkansas                4044
California              4044
Colorado                4044
Connecticut             4044
Delaware                4044
District of Columbia    4044
Florida                 4044
Georgia                 4044
Hawaii                  4044
Idaho                   4044
Illinois                4044
Indiana                 4044
Iowa                    4044
Kansas                  4043
Kentucky                4043
Louisiana               4043
Maine                   4043
Maryland                4043
Massachusetts           4043
Michigan                4043
Minnesota               4043
Mississippi             4043
Missouri                4043
Montana                 4043
Nebraska                4043
Nevada                  4043
New Hampshire           4043
New Jersey              4043
New Mexico              4043
New York                4043
North Carolina          4043
North Dakota  

All states are accounted for. There are no mispellings or duplicates. 

# Step 5. Fundamental data quality and consistency checks

In [29]:
# Check for missing values 
df_customers.isnull().sum()

user_id                  0
first_name           11259
last_name                0
gender                   0
state                    0
age                      0
date_joined              0
nbr_of_dependents        0
fam_status               0
income                   0
dtype: int64

There are 11,259 missing values in the 'first_name' column. There are no other missing values in other columns. 

In [31]:
## create a subset of missing values 
df_customers_nan = df_customers[df_customers['first_name'].isnull()== True] 

In [32]:
# View subset 
df_customers_nan.head(25)

user_id first_name   last_name  gender           state  age date_joined  \
53    76659        NaN     Gilbert    Male        Colorado   26    1/1/2017   
73    13738        NaN       Frost  Female       Louisiana   39    1/1/2017   
82    89996        NaN      Dawson  Female          Oregon   52    1/1/2017   
99    96166        NaN     Oconnor    Male        Oklahoma   51    1/1/2017   
105   29778        NaN      Dawson  Female            Utah   63    1/1/2017   
128    8562        NaN     Oconnor    Male            Utah   46    1/1/2017   
140  149267        NaN  Hutchinson    Male  South Carolina   20    1/1/2017   
149   82632        NaN         Orr    Male          Hawaii   61    1/1/2017   
155  172331        NaN  Williamson  Female          Alaska   27    1/1/2017   
236  182963        NaN   Nicholson  Female      New Mexico   58    1/2/2017   
241   49727        NaN       Moran    Male          Alaska   53    1/2/2017   
265   63554        NaN  Williamson  Female        Delaware   49    1/2/2017   
275  199686        NaN     Hampton  Female         Montana   21    1/2/2017   
286   36684        NaN    Copeland  Female         Indiana   73    1/2/2017   
294  123632        NaN       Dixon  Female        Delaware   24    1/2/2017   
338   92761        NaN       Floyd  Female       Tennessee   29    1/3/2017   
346  112064        NaN     Dejesus    Male          Kansas   48    1/3/2017   
351  151181        NaN     Oconnor    Male          Kansas   43    1/3/2017   
398   23895        NaN      Patton    Male        New York   56    1/3/2017   
407  169779        NaN     Miranda    Male         Wyoming   32    1/3/2017   
414   39945        NaN     Fischer    Male          Hawaii   53    1/3/2017   
424    4515        NaN     Gilbert    Male        New York   25    1/3/2017   
456   99868        NaN        Bean  Female         Florida   55    1/3/2017   
457    7804        NaN   Nicholson  Female         Alabama   28    1/3/2017   
467   20786        NaN      Davila  Female    North Dakota   35    1/3/2017   

     nbr_of_dependents fam_status  income  
53                   2    married   41709  
73                   0     single   82518  
82                   3    married  117099  
99                   1    married  155673  
105                  3    married  151819  
128                  1    married  134898  
140                  0     single   86778  
149                  1    married  118130  
155                  0     single   55047  
236                  1    married  163391  
241                  1    married  101623  
265                  3    married   30236  
275                  0     single   56294  
286                  2    married  103261  
294                  1    married   36790  
338                  3    married   97830  
346                  2    married   58260  
351                  3    married  128893  
398                  3    married   40763  
407                  2    married   56094  
414                  2    married  103090  
424                  1    married   28857  
456                  1    married  103489  
457                  0     single   32352  
467                  3    married   74768

In [33]:
# Check descriptive stats 
df_customers.describe()

age  nbr_of_dependents         income
count  206209.000000      206209.000000  206209.000000
mean       49.501646           1.499823   94632.852548
std        18.480962           1.118433   42473.786988
min        18.000000           0.000000   25903.000000
25%        33.000000           0.000000   59874.000000
50%        49.000000           1.000000   93547.000000
75%        66.000000           3.000000  124244.000000
max        81.000000           3.000000  593901.000000

In [34]:
# Check shape 
df_customers.shape

(206209, 10)

The missing values in the 'first_name' column will be left alone since everything else is accounted (ex: last names, gender, age, state). No columns need to be dropped at this point as well.  

In [35]:
# Check for duplicates 
df_customers_dups = df_customers[df_customers.duplicated()]

In [36]:
df_customers_dups

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, nbr_of_dependents, fam_status, income]
Index: []

There are no duplicates in the dataframe. 

The df_customers dataframe is wrangled and ready to be merged with the Instacart data. First, I'll export the df_customers dataframe to save it. 

In [38]:
# Export dataframe as a pickle file 
df_customers.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'customers_wrangled.pkl'))

In [39]:
# Import pickle files as dataframes to prepare for the merge between data sets 
df_ords_prods = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_flagged.pkl'))

In [40]:
df_cust = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'customers_wrangled.pkl'))

# Step 7. Combine data sets 

In [41]:
# Check shape of dataframes 
df_ords_prods.shape

(32404859, 25)

In [42]:
df_cust.shape

(206209, 10)

df_cust is significially smaller than df_ords_prods.

In [43]:
# Check descriptive stats of dataframes 
df_ords_prods.describe()

order_id       user_id  order_number  order_day_of_week  \
count  3.240486e+07  3.240486e+07  3.240486e+07       3.240486e+07   
mean   1.710745e+06  1.029372e+05  1.714230e+01       2.738867e+00   
std    9.872988e+05  5.946610e+04  1.753532e+01       2.090077e+00   
min    2.000000e+00  1.000000e+00  1.000000e+00       0.000000e+00   
25%    8.559470e+05  5.142200e+04  5.000000e+00       1.000000e+00   
50%    1.711049e+06  1.026160e+05  1.100000e+01       3.000000e+00   
75%    2.565499e+06  1.543890e+05  2.400000e+01       5.000000e+00   
max    3.421083e+06  2.062090e+05  9.900000e+01       6.000000e+00   

       order_hour_of_day  days_since_prior_order    product_id  \
count       3.240486e+07            3.032876e+07  3.240486e+07   
mean        1.342515e+01            1.110408e+01  2.559866e+04   
std         4.246380e+00            8.779064e+00  1.408400e+04   
min         0.000000e+00            0.000000e+00  1.000000e+00   
25%         1.000000e+01            5.000000e+00  1.354400e+04   
50%         1.300000e+01            8.000000e+00  2.530200e+04   
75%         1.600000e+01            1.500000e+01  3.794700e+04   
max         2.300000e+01            3.000000e+01  4.968800e+04   

       add_to_cart_order     reordered      aisle_id  department_id  \
count       3.240486e+07  3.240486e+07  3.240486e+07   3.240486e+07   
mean        8.352547e+00  5.895873e-01  7.119612e+01   9.919792e+00   
std         7.127071e+00  4.919087e-01  3.821139e+01   6.281485e+00   
min         1.000000e+00  0.000000e+00  1.000000e+00   1.000000e+00   
25%         3.000000e+00  0.000000e+00  3.100000e+01   4.000000e+00   
50%         6.000000e+00  1.000000e+00  8.300000e+01   9.000000e+00   
75%         1.100000e+01  1.000000e+00  1.070000e+02   1.600000e+01   
max         1.450000e+02  1.000000e+00  1.340000e+02   2.100000e+01   

             prices     max_order  spending_habits  ordering_habits  
count  3.240486e+07  3.240486e+07     3.240486e+07     3.240485e+07  
mean   1.198023e+01  3.305217e+01     1.198023e+01     1.039776e+01  
std    4.956554e+02  2.515525e+01     8.324227e+01     7.131754e+00  
min    1.000000e+00  1.000000e+00     1.000000e+00     0.000000e+00  
25%    4.200000e+00  1.300000e+01     7.387298e+00     6.000000e+00  
50%    7.400000e+00  2.600000e+01     7.824786e+00     8.000000e+00  
75%    1.130000e+01  4.700000e+01     8.254023e+00     1.300000e+01  
max    9.999900e+04  9.900000e+01     2.500542e+04     3.000000e+01

In [44]:
df_ords_prods.info() 

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 25 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   eval_set                object  
 3   order_number            int64   
 4   order_day_of_week       int64   
 5   order_hour_of_day       int64   
 6   days_since_prior_order  float64 
 7   product_id              int64   
 8   add_to_cart_order       int64   
 9   reordered               int64   
 10  _merge                  category
 11  product_name            object  
 12  aisle_id                int64   
 13  department_id           int64   
 14  prices                  float64 
 15  price_range_loc         object  
 16  busiest_day             object  
 17  busiest_days            object  
 18  busiest_period_of_day   object  
 19  max_order               int64   
 20  loyalty_flag            object  
 21  spendi

In [45]:
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   user_id            206209 non-null  object
 1   first_name         194950 non-null  object
 2   last_name          206209 non-null  object
 3   gender             206209 non-null  object
 4   state              206209 non-null  object
 5   age                206209 non-null  int64 
 6   date_joined        206209 non-null  object
 7   nbr_of_dependents  206209 non-null  int64 
 8   fam_status         206209 non-null  object
 9   income             206209 non-null  int64 
dtypes: int64(3), object(7)
memory usage: 15.7+ MB


The key column that can be used to merge both data sets is the 'user_id' column. Before the merge, the datatype for 'user_id' in df_ords_prods need to be updated as well as 'order_id', 'product_id', 'aisle_id', 'department_id'. 

In [46]:
# Change datatype in multiple columns 
df_ords_prods['user_id'] = df_ords_prods['user_id'].astype('str')

In [49]:
df_ords_prods['order_id'] = df_ords_prods['order_id'].astype('str')

In [50]:
df_ords_prods['product_id'] = df_ords_prods['product_id'].astype('str')

In [51]:
df_ords_prods['aisle_id'] = df_ords_prods['aisle_id'].astype('str')

In [52]:
df_ords_prods['department_id'] = df_ords_prods['department_id'].astype('str')

In [53]:
# Check dataframe 
df_ords_prods.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 25 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                object  
 1   user_id                 object  
 2   eval_set                object  
 3   order_number            int64   
 4   order_day_of_week       int64   
 5   order_hour_of_day       int64   
 6   days_since_prior_order  float64 
 7   product_id              object  
 8   add_to_cart_order       int64   
 9   reordered               int64   
 10  _merge                  category
 11  product_name            object  
 12  aisle_id                object  
 13  department_id           object  
 14  prices                  float64 
 15  price_range_loc         object  
 16  busiest_day             object  
 17  busiest_days            object  
 18  busiest_period_of_day   object  
 19  max_order               int64   
 20  loyalty_flag            object  
 21  spendi

Data types of all columns are all checked out. 

In [54]:
# View dataframes 
df_ords_prods.head()

order_id user_id eval_set  order_number  order_day_of_week  \
0  2539329       1    prior             1                  2   
1  2398795       1    prior             2                  3   
2   473747       1    prior             3                  3   
3  2254736       1    prior             4                  4   
4   431534       1    prior             5                  4   

   order_hour_of_day  days_since_prior_order product_id  add_to_cart_order  \
0                  8                     NaN        196                  1   
1                  7                    15.0        196                  1   
2                 12                    21.0        196                  1   
3                  7                    29.0        196                  1   
4                 15                    28.0        196                  1   

   reordered  ...    price_range_loc     busiest_day  busiest_days  \
0          0  ...  Mid-range product  Regularly busy   Normal days   
1          1  ...  Mid-range product  Regularly busy  Slowest days   
2          1  ...  Mid-range product  Regularly busy  Slowest days   
3          1  ...  Mid-range product      Least busy  Slowest days   
4          1  ...  Mid-range product      Least busy  Slowest days   

  busiest_period_of_day  max_order  loyalty_flag spending_habits  \
0        Average orders         10  New customer        6.367797   
1         Fewest orders         10  New customer        6.367797   
2           Most orders         10  New customer        6.367797   
3         Fewest orders         10  New customer        6.367797   
4           Most orders         10  New customer        6.367797   

  spending_flag ordering_habits        order_freq_flag  
0   Low spender            20.5  Non-frequent customer  
1   Low spender            20.5  Non-frequent customer  
2   Low spender            20.5  Non-frequent customer  
3   Low spender            20.5  Non-frequent customer  
4   Low spender            20.5  Non-frequent customer  

[5 rows x 25 columns]

In [55]:
# Command pandas not to assign any options regarding the max number of columns to display 
pd.options.display.max_columns = None 

In [56]:
df_ords_prods.head()

order_id user_id eval_set  order_number  order_day_of_week  \
0  2539329       1    prior             1                  2   
1  2398795       1    prior             2                  3   
2   473747       1    prior             3                  3   
3  2254736       1    prior             4                  4   
4   431534       1    prior             5                  4   

   order_hour_of_day  days_since_prior_order product_id  add_to_cart_order  \
0                  8                     NaN        196                  1   
1                  7                    15.0        196                  1   
2                 12                    21.0        196                  1   
3                  7                    29.0        196                  1   
4                 15                    28.0        196                  1   

   reordered _merge product_name aisle_id department_id  prices  \
0          0   both         Soda       77             7     9.0   
1          1   both         Soda       77             7     9.0   
2          1   both         Soda       77             7     9.0   
3          1   both         Soda       77             7     9.0   
4          1   both         Soda       77             7     9.0   

     price_range_loc     busiest_day  busiest_days busiest_period_of_day  \
0  Mid-range product  Regularly busy   Normal days        Average orders   
1  Mid-range product  Regularly busy  Slowest days         Fewest orders   
2  Mid-range product  Regularly busy  Slowest days           Most orders   
3  Mid-range product      Least busy  Slowest days         Fewest orders   
4  Mid-range product      Least busy  Slowest days           Most orders   

   max_order  loyalty_flag  spending_habits spending_flag  ordering_habits  \
0         10  New customer         6.367797   Low spender             20.5   
1         10  New customer         6.367797   Low spender             20.5   
2         10  New customer         6.367797   Low spender             20.5   
3         10  New customer         6.367797   Low spender             20.5   
4         10  New customer         6.367797   Low spender             20.5   

         order_freq_flag  
0  Non-frequent customer  
1  Non-frequent customer  
2  Non-frequent customer  
3  Non-frequent customer  
4  Non-frequent customer

All columns are in view. 

In [57]:
df_cust.head()

user_id first_name last_name  gender       state  age date_joined  \
0   26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1   33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2   65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3  125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4  130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   nbr_of_dependents fam_status  income  
0                  3    married  165665  
1                  0     single   59285  
2                  2    married   99568  
3                  0     single   42049  
4                  1    married   40374

As mentioned earlier, df_cust is signficially smaller than df_ords_prods. To prep for a merge, I find a shared column to use as a key; "user_id" columns is a full match in both dataframes. The default type for a join is INNER, which means the resulting data set will only contain observations included in both input data sets. 

In [60]:
# Merge dataframes 
df_merge = df_ords_prods.merge(df_cust, on = 'user_id')

In [61]:
# Check merge 
df_merge['_merge'].value_counts()

both          32404859
right_only           0
left_only            0
Name: _merge, dtype: int64

In [62]:
# Merge dataframes using INNER JOIN 
df_merge_1 = df_ords_prods.merge(df_cust, on = 'user_id', how = 'inner')

In [63]:
# Check merge 
df_merge_1['_merge'].value_counts()

both          32404859
right_only           0
left_only            0
Name: _merge, dtype: int64

CareerFoundry notified me that there will be a full merge rate between both data sets so I won't need to double check for a merge rate between both data sets. If I did, all I would need to do is to create another merge and use the OUTER JOIN (how = 'outer') argument in the code like this: <br> 
<br><b>df_merge_2 = df_ords_prods.merge(df_cust, on = 'user_id', how = 'outer')</b>

In [65]:
# Check new dataframe 
df_merge_1.head(100)

order_id user_id eval_set  order_number  order_day_of_week  \
0   2539329       1    prior             1                  2   
1   2398795       1    prior             2                  3   
2    473747       1    prior             3                  3   
3   2254736       1    prior             4                  4   
4    431534       1    prior             5                  4   
..      ...     ...      ...           ...                ...   
95  3317979      15    prior             5                  4   
96  2685110      15    prior             7                  1   
97   887727      15    prior             9                  2   
98  2600170      15    prior            11                  2   
99  2332087      15    prior            13                  1   

    order_hour_of_day  days_since_prior_order product_id  add_to_cart_order  \
0                   8                     NaN        196                  1   
1                   7                    15.0        196                  1   
2                  12                    21.0        196                  1   
3                   7                    29.0        196                  1   
4                  15                    28.0        196                  1   
..                ...                     ...        ...                ...   
95                 15                    17.0      14715                  1   
96                 11                    17.0      14715                  3   
97                 13                     7.0      14715                  1   
98                  9                    14.0      14715                  1   
99                 10                    14.0      14715                  3   

    reordered _merge   product_name aisle_id department_id  prices  \
0           0   both           Soda       77             7     9.0   
1           1   both           Soda       77             7     9.0   
2           1   both           Soda       77             7     9.0   
3           1   both           Soda       77             7     9.0   
4           1   both           Soda       77             7     9.0   
..        ...    ...            ...      ...           ...     ...   
95          1   both  Coconut Water       98             7     4.0   
96          1   both  Coconut Water       98             7     4.0   
97          1   both  Coconut Water       98             7     4.0   
98          1   both  Coconut Water       98             7     4.0   
99          1   both  Coconut Water       98             7     4.0   

      price_range_loc     busiest_day  busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy   Normal days        Average orders   
1   Mid-range product  Regularly busy  Slowest days         Fewest orders   
2   Mid-range product  Regularly busy  Slowest days           Most orders   
3   Mid-range product      Least busy  Slowest days         Fewest orders   
4   Mid-range product      Least busy  Slowest days           Most orders   
..                ...             ...           ...                   ...   
95  Low-range product      Least busy  Slowest days           Most orders   
96  Low-range product  Regularly busy  Busiest days           Most orders   
97  Low-range product  Regularly busy   Normal days           Most orders   
98  Low-range product  Regularly busy   Normal days           Most orders   
99  Low-range product  Regularly busy  Busiest days           Most orders   

    max_order      loyalty_flag  spending_habits spending_flag  \
0          10      New customer         6.367797   Low spender   
1          10      New customer         6.367797   Low spender   
2          10      New customer         6.367797   Low spender   
3          10      New customer         6.367797   Low spender   
4          10      New customer         6.367797   Low spender   
..        ...               ...              ...           ...   
95         22  Regular customer         3.980

In [66]:
# Command pandas not to assign any options regarding the max number of rows to display 
pd.options.display.max_rows = None 

In [67]:
df_merge_1.head(100)

order_id user_id eval_set  order_number  order_day_of_week  \
0   2539329       1    prior             1                  2   
1   2398795       1    prior             2                  3   
2    473747       1    prior             3                  3   
3   2254736       1    prior             4                  4   
4    431534       1    prior             5                  4   
5   3367565       1    prior             6                  2   
6    550135       1    prior             7                  1   
7   3108588       1    prior             8                  1   
8   2295261       1    prior             9                  1   
9   2550362       1    prior            10                  4   
10  2539329       1    prior             1                  2   
11  2539329       1    prior             1                  2   
12  2398795       1    prior             2                  3   
13   473747       1    prior             3                  3   
14  2254736       1    prior             4                  4   
15   431534       1    prior             5                  4   
16  3367565       1    prior             6                  2   
17   550135       1    prior             7                  1   
18  3108588       1    prior             8                  1   
19  2295261       1    prior             9                  1   
20  2550362       1    prior            10                  4   
21  2539329       1    prior             1                  2   
22  2398795       1    prior             2                  3   
23  2539329       1    prior             1                  2   
24  2254736       1    prior             4                  4   
25  2398795       1    prior             2                  3   
26   473747       1    prior             3                  3   
27  2254736       1    prior             4                  4   
28   431534       1    prior             5                  4   
29  3367565       1    prior             6                  2   
30   550135       1    prior             7                  1   
31  3108588       1    prior             8                  1   
32  2295261       1    prior             9                  1   
33  2550362       1    prior            10                  4   
34  2398795       1    prior             2                  3   
35   431534       1    prior             5                  4   
36  2398795       1    prior             2                  3   
37   550135       1    prior             7                  1   
38  2550362       1    prior            10                  4   
39   473747       1    prior             3                  3   
40  2254736       1    prior             4                  4   
41   431534       1    prior             5                  4   
42  3367565       1    prior             6                  2   
43   550135       1    prior             7                  1   
44  3108588       1    prior             8                  1   
45  2295261       1    prior             9                  1   
46  2550362       1    prior            10                  4   
47   473747       1    prior             3                  3   
48   431534       1    prior             5                  4   
49   431534       1    prior             5                  4   
50   431534       1    prior             5                  4   
51  3108588       1    prior             8                  1   
52  2295261       1    prior             9                  1   
53  2550362       1    prior            10                  4   
54  3108588       1    prior             8                  1   
55  2295261       1    prior             9                  1   
56  2550362       1    prior            10                  4   
57  2550362       1    prior            10                  4   
58  2550362       1    prior            10                  4   
59  2968173      15    prior            15                  1   
60  1870022      15    prior          

In [68]:
# Check shape 
df_merge_1.shape

(32404859, 34)

There are 32,404,859 rows in this dataframe with 34 columns. This is a large dataframe. Both data sets are fully merged. 

# Steps 7, 8, 9 <br>

The notebook contains logical titles, section headings, and descriptive code comments.

In [71]:
# Export dataframe 
df_merge_1.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_cust.pkl'))